Added support of NANOO fp8.

wenchenvincent opened the pull request:
# What does this PR do?
This PR adds support of fp8 dot op for NANOO fp8 data formats (an alternative genre to the OCP fp8 data formats, which is used by NVIDIA GPU).

There are several different genres of fp8 formats used by different HW vendors. Two popular genres include
- OCP fp8, which is used natively on NVIDIA H100
- NANOO fp8, which is used natively on AMD MI300 and Graphcore HW.

These two genres of fp8 formats work very similarly. This PR is to enable support of NANOO fp8 as it is also now supported in JAX and XLA. It would enable usage of fp8 dot op on AMD MI300 GPU.

References:
- OCP fp8 paper: https://arxiv.org/abs/2209.05433
- NANOO fp8 paper: https://arxiv.org/abs/2206.02915
- JAX PR: https://github.com/google/jax/pull/21376
- XLA PR: https://github.com/openxla/xla/pull/9531

wenchenvincent commented:
@levskaya I noticed that you have reviewed several PRs regarding fp8. Could you take a look at this one?

wenchenvincent commented:
@levskaya Could you kindly serve as the reviewer for this PR?

levskaya commented on flax/linen/fp8_ops.py:
```diff
@@ -259,6 +269,7 @@ def dot_general_with_precision_jvp(
 
 
 class Fp8DotGeneralOp(module.Module):
+  fp8_genre: str = 'OCP'
```
instead of indirecting through a string lookup in `get_fp8_dtypes`, can we please just make the fwd/bwd fp8 dtypes class attributes and default to existing "ocp" types, and place the new fields _after_ the already existing one
```
amax_history_length: int = 1024
e4m3_dtype: Dtype = jnp.float8_e4m3fn
e5m2_dtype: Dtype = jnp.float8_e5m2
```

(feel free to define a typing alias dtype up top:
```
from typing import Any
Dtype = Any 
```
)

wenchenvincent replied:
Done.

levskaya commented on flax/linen/fp8_ops.py:
```diff
@@ -306,18 +317,27 @@ def __call__(self, *args, **kwargs):
+  fp8_genre: str = 'OCP'
+
+class NANOOFp8DotGeneralOp(Fp8DotGeneralOp):
+  fp8_genre: str = 'NANOO'
```
e.g. this would now be:
```
class NANOOFp8DotGeneralOp(Fp8DotGeneralOp):
  e4m3_dtype: Dtype = jnp.float8_e4m3fnuz
  e5m2_dtype: Dtype = jnp.float8_e5m2nuz
```

wenchenvincent replied:
Done.

levskaya commented on flax/linen/fp8_ops.py:
```diff
@@ -306,18 +317,27 @@ def __call__(self, *args, **kwargs):
 
     return y  # type: ignore
+
+class OCPFp8DotGeneralOp(Fp8DotGeneralOp):
```
as said above, I'd drop this new symbol as we don't want to break existing api.

wenchenvincent replied:
Done.

levskaya reviewed (commented):
Sorry for the delay!

Looks OK, but a few requests to simplify class configuration and to not break existing names in public API.

Try resubmitting for tests after fixing that, the failure before was from a transient unrelated breakage.

wenchenvincent commented on flax/linen/__init__.py:
```diff
@@ -86,7 +86,8 @@
 )
 from .batch_apply import BatchApply as BatchApply
 from .combinators import Sequential as Sequential
-from .fp8_ops import Fp8DotGeneralOp as Fp8DotGeneralOp
```
Thanks for the review. I have reverted the change.

wenchenvincent commented:
@levskaya Thanks for the review. I have updated the PR to address the concerns. Could you take a look at the updates?

levskaya reviewed (approved):
Thanks for the fixes!  We may need to do some tiny rebasing of simple things as the codebase just migrated to a python minver of 3.10.

wenchenvincent commented:
> Thanks for the fixes! We may need to do some tiny rebasing of simple things as the codebase just migrated to a python minver of 3.10.

Thanks! Do you need me to rebase it to the tip of the  tree?

levskaya commented:
Yes to tip as of today should have the 3.10 minver updates.   Also, I'm seeing this failure in the tests:
```
FAILED tests/linen/linen_test.py::Fp8Test::test_fp8_meta_dtype0 - TypeError: missing a required argument: 'amax_history'
FAILED tests/linen/linen_test.py::Fp8Test::test_fp8_meta_dtype1 - TypeError: missing a required argument: 'amax_history'
```
could you fix that?

wenchenvincent commented:
> Yes to tip as of today should have the 3.10 minver updates. Also, I'm seeing this failure in the tests:
> 
> ```
> FAILED tests/linen/linen_test.py::Fp8Test::test_fp8_meta_dtype0 - TypeError: missing a required argument: 'amax_history'
> FAILED tests/linen/linen_test.py::Fp8Test::test_fp8_meta_dtype1 - TypeError: missing a required argument: 'amax_history'
> ```
> 
> could you fix that?

Sorry I missed this test.

I just rebased and fixed this test.

levskaya commented on tests/linen/linen_test.py:
```diff
@@ -1421,12 +1450,10 @@ def body_fun(carry, _):
+    np.testing.assert_allclose(new_sf, [3. / e4m3_max])
 
 
-if __name__ == '__main__':
```
uh, don't delete this.  many test runners won't run tests without it.

wenchenvincent replied:
Sorry, it was accidentally removed. Didn't mean to. I just restored it.

levskaya commented on flax/linen/fp8_ops.py:
```diff
@@ -117,9 +120,19 @@ def __repr__(self) -> str:
+                       jnp.float8_e4m3fnuz, jnp.float8_e5m2fnuz)
   return jnp.finfo(fp8_dtype).max.astype(out_dtype)
 
+def get_fp8_dtypes(fp8_genre):
```
this function seems to only be used by tests, can we just move it into the test file?

wenchenvincent replied:
Done.

levskaya reviewed (commented):
Hey sorry, after importing, I just noticed two more things that need to be fixed.
